### PR TITLE
updated markdown in 01 notebook up to fit function, and plot_training…

### PR DESCRIPTION
**Pull request by karineip**
…_losses function in model.py

Based off Casey's PR1 merged into staging, 
- updated 01 notebook for fit section with markdown based on Patrick's feedback on closed PR #545, 
- added plot_training_losses function

Request feedback on:
- plotting losses, what's should be discussed in markdown

### Checklist:

- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`
- [ ] I will squash merge this PR into `staging`

### CODE REVIEW

**Comment by PatrickBue** [comment]
@karineip Strangely this PR does not seem to contain any code edits. Please inspect the "files changes" tab, including all individual code/text changes, every time before adding other people to the code review to make sure all is as expected.

**Comment by karineip** [comment]
Apologies, I got mixed up when changing branches :( I updated the files in the recent commit.

File: utils_cv/tracking/model.py [hunk 1]
```diff
@@ -177,15 +177,45 @@ def fit(
         trainer.set_device(
             self.opt.gpus, self.opt.chunk_sizes, self.opt.device
         )
-
+        
+        # initialize loss vars
+        self.losses_dict = {key: [] for key in ['loss', 'hm_loss', 'wh_loss', 'off_loss', 'id_loss']}
```
Comment (PatrickBue) [inline utils_cv/tracking/model.py:182]: [minor] I believe you can use 'defaultdict" here which does not need to be initialized with an empty list.
   from collections import defaultdict
   self.loss_dict = defaultdict()

File: utils_cv/tracking/model.py [hunk 2]
```diff
@@ -177,15 +177,45 @@ def fit(
...
             log_dict_train, _ = trainer.train(epoch, train_loader)
             if epoch in self.opt.lr_step:
                 lr = self.opt.lr * (0.1 ** (self.opt.lr_step.index(epoch) + 1))
                 for param_group in optimizer.param_groups:
                     param_group["lr"] = lr
+                    
+            # store losses in each epoch            
+            self.epochs.append(epoch)
+            for k, v in log_dict_train.items():
+                if k in ['loss', 'hm_loss', 'wh_loss', 'off_loss', 'id_loss']:
```
Comment (PatrickBue) [inline utils_cv/tracking/model.py:198]: [minor] This line when using defaultdict is probably not needed anymore then

File: utils_cv/tracking/model.py [hunk 3]
```diff
@@ -177,15 +177,45 @@ def fit(
...
+                    self.losses_dict[k].append(v)
+
+    def plot_training_losses(self, figsize: Tuple[int, int] = (10, 5))->None:
```
Comment (PatrickBue) [inline utils_cv/tracking/model.py:201]: Please for function comments use the same syntax as everywhere in the code, as e.g. the fit method, ie:
        """
        <text description>
       
        Args:
            figsize: <text>
        """

File: scenarios/tracking/01_training_introduction.ipynb [hunk 4]
```diff
@@ -13,19 +13,36 @@
    "cell_type": "markdown",
```
Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <ul><li>Nice summary. Minor suggestions below:</li><li>Would rephrase sentence to: In particular, we will use <a href="https://github.com/ifzhang/FairMOT" target="_blank">FairMOT</a> which is a state-of-the-art tracking model with high accuracy and inference speed.</li><li>Then delete the "which detects objects and learns their re-ID features"</li><li>Remove sentence "We also show how to save and load the trained model for inference on a second video.  "</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Please sort imports first by type (block for native imports, block for third party imports, block for our utils_cv imports) and then within the blocks alphabetically</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>There are some variables in here which seem are not used. Please remove and only keep the ones that are really necessary.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <ul><li>Please remove TODO statements so that this is clean and can go into staging</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>This really creates an object of TrackingDataset, hence update sentence to "To load the data, we create a <code>TrackingDataset</code>  object which knows how to read our images and annotations."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <ul><li>Minor: no need to point out what the main parameters are since they are in the fuction call anyway.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <ul><li>minor: we don't generate losses, rather while training losses are being computed</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Comment by karineip** [comment]
I've pushed the updated files, with updates per @Patrick's feedback. I think the workflow is to close this PR, but incorporate the changes into a new branch off Casey's next PR 553 (which includes predict) once it's merged into staging. @caseyhong please correct me if wrong.

File: utils_cv/tracking/model.py [hunk 5]
```diff
@@ -177,15 +178,51 @@ def fit(
...
```
Comment (PatrickBue) [inline utils_cv/tracking/model.py:202]: Please move this function to plot.py

File: utils_cv/tracking/model.py [hunk 6]
```diff
@@ -177,15 +178,51 @@ def fit(
...
+            print("Training in epoch ", epoch,"...")
             mark = epoch if self.opt.save_all else "last"
             log_dict_train, _ = trainer.train(epoch, train_loader)
             if epoch in self.opt.lr_step:
                 lr = self.opt.lr * (0.1 ** (self.opt.lr_step.index(epoch) + 1))
                 for param_group in optimizer.param_groups:
                     param_group["lr"] = lr
+                    
+            # store losses in each epoch            
+            self.epochs.append(epoch)
```
Comment (PatrickBue) [inline utils_cv/tracking/model.py:197]: Is this self.epochs used anywhere (except in the plot function where it might not be needed)

File: scenarios/tracking/01_training_introduction.ipynb (hunk 4 above)
Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Please remove old code everywhere, ie here:</p><p>#from utils_cv.multi_object_tracking.display_with_bb import convert_trackingbboxes_video</p><p>#from utils_cv.multi_object_tracking.file_format import convert_vott_MOTxywh </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (PatrickBue) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>DATA_PATH_TRAIN can be removed here since it is already specified above</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

**Comment by PatrickBue** [comment]
@karineip left a couple of minor suggestions. One addressed this can go into staging. Note that you'll have merge conflicts since the notebook has been updated. Likely it'll be easiest to re-add your text into casey's new notebook.

File: scenarios/tracking/01_training_introduction.ipynb (hunk 4 above)
Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>How about:</p><p>using <s>torchvision</s> FairMOT</p><p><s>In particular, we will use </s>FairMOT <s>which</s> is a state-of-the-art tracking model with high accuracy and inference speed.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Could you remove the commented lines? Also run black formatting on the notebook cells</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Nit: a small training <strong>dataset</strong></p><p> </p><p>Do we need to have the last two sentences on download/unzip, and setting the path?  These steps are pretty self-explanatory in the code. </p><p>Also the sentence on path is a little confusing, as it could make the reader think of the PATH environment variable.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>To load the data, we create a <code>TrackingDataset</code> object <s>which knows how to read our images and annotations. </s>with the path to our training dataset and desired batch size.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>How about:</p><p><code>TrackingLearner</code> defaults to FairMOT's baseline tracking model, which can be downloaded [here] &lt;link to gdrive&gt;. This has been pretrained on pedestrian data, such as those found in the MOT Challenge datasets. In order to track the fridge objects in our use-case, we must finetune the model on our own dataset.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Seems slightly superfluous, the fit() call is self-explanatory  </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (caseyhong) [inline scenarios/tracking/01_training_introduction.ipynb:13]: <p>Perhaps some of this material can be moved to the FAQ, and be made more succinct here:</p><p> </p><p><s>We can now</s> <strong>Here we</strong> visualize the training losses computed in each training epoch<s>, and see how the model improves with training. We want to run the training for an appropriate </s><code><s>num_epochs</s></code><s> and </s><code><s>lr</s></code><s> (to be fine-tuned by the user) that produces a loss-curve that tails off. The loss-curve for our training is as follows:</s></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/computervision-recipes/pull/558/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

File: utils_cv/tracking/model.py [hunk 7]
```diff
@@ -177,15 +178,51 @@ def fit(
...
         )
-
+        
+        # initialize loss vars
+        self.losses_dict = defaultdict(list)
+        self.epochs = []
+        
         # training loop
         for epoch in range(self.start_epoch + 1, self.opt.num_epochs + 1):
+            print("Training in epoch ", epoch,"...")
```
Comment (caseyhong) [inline utils_cv/tracking/model.py:188]: Please remove and use the logging statements in the training loop in staging.

File: utils_cv/tracking/model.py [hunk 8]
```diff
@@ -177,15 +178,51 @@ def fit(
...
+        Plots training loss from calling `fit`  
+        
+        Args:
+            figsize (optional): width and height wanted for figure of training-loss plot
+        
+        '''
+        fig = plt.figure(figsize=figsize)
+        ax1 = fig.add_subplot(1, 1, 1)
+
+        #ax1.set_xlim([0, self.epochs - 1])
```
Comment (caseyhong) [inline utils_cv/tracking/model.py:213]: Remove commented line

File: utils_cv/tracking/model.py [hunk 9]
```diff
@@ -177,15 +178,51 @@ def fit(
...
+        ax1.set_xlabel("epochs")
+        ax1.set_ylabel("losses")
+        ax1.plot(self.losses_dict['loss'], c="r", label='loss')
+        ax1.plot(self.losses_dict['hm_loss'], c="y", label='hm_loss')
+        ax1.plot(self.losses_dict['wh_loss'], c="g", label='wh_loss')
+        ax1.plot(self.losses_dict['off_loss'], c="b", label='off_loss')
+        ax1.plot(self.losses_dict['id_loss'], c="m", label='id_loss')
+
+        plt.legend(loc='upper right')
+        fig.suptitle("Loss over epochs during training")
```
Comment (caseyhong) [inline utils_cv/tracking/model.py:224]: Nit but how about "Training loss over epochs"

File: utils_cv/tracking/model.py (hunk 5 above)
Comment (caseyhong) [reply utils_cv/tracking/model.py:202]: FWIW in the detection scenario `plot_precision_loss_curve` is in the model class

Comment (karineip) [reply utils_cv/tracking/model.py:202]: @PatrickBue  in the OD scenario,  `plot_precision_loss_curve` is in the model class. Should I still move this function to plot.py?

Comment (karineip) [reply utils_cv/tracking/model.py:202]: If I move `plot_training_losses` to plot.py, should I change `fit` to return `self.losses_dict` , and pass it on in the notebook with, i.e. `losses_dict = fit(...)` and `plot_training_losses(losses_dict)` ?

Comment (PatrickBue) [reply utils_cv/tracking/model.py:202]: Good point, for object detection the plot_ function is in models.py: https://github.com/microsoft/computervision-recipes/blob/master/utils_cv/detection/model.py#L549
In that case this function can stay here. My comments about self.epochs stays the same, I don't think it's needed.
